### ISSUE TITLE
JabRef breaks hard links to .bib files

### ISSUE THREAD

**Issue by sparusaurata** [JabRef/jabref#7718]
I use JabRef version 5.2 on Linux Mint 20.1.
When I open a `.bib` file and modify it with JabRef, its inode number is modified and the hard links are broken. An example:
```
$ ln these.bib Simulation\ βinf

$ ls -i these.bib
9440991 these.bib

$ ls -i Simulation\ βinf/these.bib
9440991 'Simulation βinf/these.bib'

# Open JabRef and modify these.bib

$ ls -li these.bib                
9437275 -rw-rw-r-- 1 remy remy 26060 mai    8 16:51 these.bib

$ ls -li Simulation\ βinf/these.bib      
9440991 -rw-rw-r-- 1 remy remy 26060 mai    8 16:38 'Simulation βinf/these.bib'
```

**Comment by Siedlerchr** [comment]
This is correct in so far that JabRef writes all content to a temporary file first and then replaces the existing file in an atomic move where possible.

**Comment by sparusaurata** [comment]
And would it be conceivable to do it some other way (naive question)?
(In my case using hard links is necessary because I work in a git repo, and symbolic links would not be pushed correctly to remotes.)

**Comment by Siedlerchr** [comment]
This somehow refs https://github.com/JabRef/jabref/pull/6694 There is a lengthy discussion. 
I have been working on this saving already and I am experimenting if this works fine if I directly write to the bib file.

May I ask why you need a hard link? Why don't you edit the file directly in the repo?

**Comment by sparusaurata** [comment]
> May I ask why you need a hard link? Why don't you edit the file directly in the repo?
The main `.bib` file, located in some root folder, is used by several projects (and associated git repos) located in subfolders. That's why I created a (hard) link to the main bibliography in each repo.

An easy workaround (that is sufficient for my projects) is to create copies instead of links, and to renew them from time to time. But in general it doesn't seem very satisfactory.

**Comment by ilippert** [comment]
I have been frustrated for years about this.

My workaround is now simplistic: I store my core library at a specific place X. And in the bibtex file I reference that file F with its complete path.

**Comment by sparusaurata** [comment]
> My workaround is now simplistic: I store my core library at a specific place X. And in the bibtex file I reference that file F with its complete path.

But this wouldn't solve my git problem, would it? (See [that comment](#issuecomment-835495739).)

**Comment by glciampaglia** [comment]
+1 --- I just stumbled on this issue. Like @sparusaurata, I also keep my bib file in a central location and link to it within individual latex projects. Normally, I use a symbolic link, but this time I have created a hard link instead because I am using Overleaf's own git repository for syncing. So neither a symbolic link, nor using a full path (as suggested by @ilippert) would work. 

I was excited about trying out this different setup because it would make it less painful to use JabRef with Overleaf projects, which now are the norm. 

(Right now when you want to edit a bib file stored in an Overleaf project you need to download it from Overleaf, edit it with JabRef, and then reupload it.)

**Comment by koppor** [comment]
This is a hard one. We need to rethink of our writing. We were pointed to different implementation approaches at  https://github.com/JabRef/jabref/pull/8750#issuecomment-1214455276.

@glciampaglia Does VisualStudio Code properly handle hard links?

**Comment by glciampaglia** [comment]
@koppor I am not sure since I am not a VSCode user myself. I am afraid I don't have technical skills to be very much of help here, but can only say that solving this issue would definitely make it possible to keep a centralized DB shared across multiple projects, as @sparusaurata mentioned.

**Comment by koppor** [comment]
For implementation:

- [ ] Write tests by calling JabRef's functionality with different settings of file links on a unix file system. This could maybe ref https://github.com/JabRef/jabref/issues/110.

**Comment by ThiloteE** [comment]
Maybe precondition for https://github.com/JabRef/jabref/issues/12350

**Comment by koppor** [comment]
It seems that one can check if a file is hard link in Java:

```java
Object nlink = Files.getAttribute(path, "unix:nlink");
if (((Integer) nlink) > 1) {
    // more than one name points at this inode
}
```

**Comment by calixtus** [comment]
Constraints
- Avoid data loss / data integrity
- Hard links working
- Performance

Current implementation:
- A new file is written, when the file is written, it replaces the existing one. 

AI analyzation of LibreOffice code:
LibreOffice already implements the relevant hybrid strategy.
It serializes the document to a temporary file, then usually replaces the original—fast and crash-safe, including for large documents. But on Unix it checks st_nlink; if it finds a hard link (or a symlink), it deliberately refuses the replace path. LibreOffice source
Its fallback is:

1. Create an internal backup of the original.
2. Truncate the original file in place.
3. Stream the completed temporary document into that same file.
4. Restore the backup if the operation fails.

That preserves the inode and hence all hard links, while using the backup as recovery protection. Fallback implementation
This is a strong model for JabRef: retain the existing AtomicFileWriter replacement path normally, but when nlink > 1, use a “backup + in-place copy” commit mode. It does sacrifice strict atomicity during the final copy, but avoids silently breaking hard links and retains a recoverable .sav backup.

**Comment by koppor** [comment]
We found the source of libre office: https://github.com/LibreOffice/core/blob/master/sfx2/source/doc/docfile.cxx#L242-L270
